R/data.table.R

```diff
@@ -266,8 +266,8 @@ replace_dot_alias = function(e) {
       with=FALSE
       if (length(av)) {
         for (..name in av) {
-          name = substring(..name, 3L)
-          if (name=="") stop("The symbol .. is invalid. The .. prefix must be followed by at least one character.")
+          name = substr(..name, 3L, nchar(..name))
```

Comment:
Writing out loud ... I suspect that passing `stop=1000000L` here instead of `nchar(..name)` would work, and avoid the allocation for the `nchar()` result.
Passing `.Machine$integer.max` might be more robust. But `substring()`'s default for `last=` is a hard coded `1000000L`. Maybe R's internal max string length is indeed `1000000L` characters, but if they ever change that they'll have to remember to update `substring`'s `last=` argument. If R's maximum string length is greater than `1000000L` then `substring`'s argument is already not correct and a bug report could be created.

So, let's see ...
```R
word = paste0(c(rep("A",1e6-2),"hello"), collapse="")
substring(word, 1e6-5)
# [1] "AAAAhe"        # unexpectedly chops "llo" off
substring(word, 1e6-5, nchar(word))
# [1] "AAAAhello"    # expected result
```

So that's a bug in R it seems. I looked at `?substring` and although the default of 10000000 is there in the definition, I don't see any text indicating this is intended behaviour.  And I see no reason that the default could not be 2^31-1 which is what the max string length is in R, iiuc.  Do you feel like reporting it @MichaelChirico? I feel like you get some traction on BugZilla and enjoy interacting there. So I'd appreciate it if you could handle that :-)

Reply:
I will take a look. My own quick investigation:

```
# works
x = strrep(" ", .Machine$integer.max-1)
# fails
x = strrep(" ", .Machine$integer.max)
```

So I agree it seems like an odd default.

Reply:
Posted here: https://stat.ethz.ch/pipermail/r-devel/2021-June/080841.html

Reply:
Great. Interested to see what their response is. I think my short example would have been better to post as-is, as the post to me on first reading dwells too much on `.Machine$integer.max` vs `.Machine$integer.max-1`. A phrase like "makes no sense" can be seen as critical and therefore risks raising hairs and inducing defensive responses. Better just to stick to code, use less English, and be constructive. But who am I to give advice, I didn't even want to engage on R-devel or Bugzilla myself because these difficulties are magnified 10x on those forums. You are braver than I.

Reply:
I see the replies now and saw that Brodie concentrated on `.Machine$integer.max` vs `.Machine$integer.max-1`. That to me is inefficient communication, as I feared. But it got there in the end with Martin getting to the root matter.

Reply:
I can guarantee you that some people (perhaps most, perhaps a few, certainly not all, but definitely some) reading that thread now think that the problem is to do with strings that are 2GB long. They therefore think that that it's a silly edge case that hardly ever comes up and folk that do have strings that are 2GB long shouldn't be doing that anyway and instead be restructuring their code.
Where in fact, the problem occurs at 1e6 characters, which is 976K.  At under 1MB, let alone GB, a string that large is relatively much more reasonable, commonplace and has relatively little to do with large servers or esoteric edge cases.
That view being formed could have been avoided by using the 1e6 example that I showed above, which is why I created it that way with that in mind. Which is easy to say in hindsight. I'm saying it now for future reference, for next time.